### ISSUE TITLE
Why is the training probability scaled by 0.5 in GaussianEncoder.get_prob_train？

### ISSUE THREAD

**Issue by Zyan20** [microsoft/DCVC#164]
Hi, thanks for the great work and open-sourcing the code!
 
In `src/models/common_model.py`, `GaussianEncoder.get_prob_train` (around line 177) computes the bin probability and then halves it before converting to bits:
 
```python
upper = _cdf2((0.5 - values) / scales)
lower = _cdf2((-0.5 - values) / scales)
prob = upper - lower
prob = torch.clamp_min(0.5 * prob, 1e-9)
```
 
Since `bits = -log2(0.5 * prob) = -log2(prob) + 1`, this charges each symbol roughly **+1 bit** relative to the true bin probability mass. Meanwhile, the quantization CDF table built in `GaussianEncoder.update()` uses the *unscaled* pmf, and the actual arithmetic coder encodes at that physical rate.
 
**Observations from my experiments** (image model, Kodak, `cvpr2026_image.pth.tar`, skip_thres both 0 and 0.15):
 
| QP | forward BPP (estimated) | compress BPP (actual bitstream) |
|----|------------------------|--------------------------------|
| 0  | 0.198 | 0.031 |
| 20 | 0.282 | 0.107 |
| 40 | 0.511 | 0.338 |
 
If I remove the `0.5 *` factor and recompute `bits_y`, the estimated BPP becomes very close to the actual bitstream BPP (e.g. QP 0: 0.207 → 0.035 vs 0.031 actual). The gap is almost entirely in `bits_y` (z uses `bit_estimator_z`, which has no such scaling), and it is *not* caused by `skip_thres`.
 
**My questions:**
 
1. What is the intended role of this `0.5` scaling? Is it (a) a deliberately conservative entropy estimate during training to leave headroom for CDF discretization / the skip mechanism, (b) only about making the reported training bpp conservative, or (c) something that actively shapes the training objective?
 
2. Since a constant multiplicative factor does not change the gradient direction of most symbols, my understanding is that it does not change the optimum of the RD objective — is that correct, or does it interact with the clamp boundary / some other mechanism in a way that matters for training dynamics?
 
Any clarification would be very helpful. Thank you!

**Comment by aviadlazar** [comment]
I think this is meant to shape the training objective because it directly affects the rate portion of the loss.

The added bit only affects probabilities that are significantly higher than 1e-9, which should encourage the entropy model to learn a distribution that tends to produce a lower probability, and avoid the added cost. This would lead to a lower entropy in general, and better compression.

Hopefully we could get a confirmation from the authors, but you could run a test that checks how the entropy during training changes with and without scaling.

**Comment by Zyan20** [comment]
Thanks for taking the time to dig into this and write it up so clearly — much appreciated!

We went back through the code and realized our measurement was mistaken. The key is that the raw CDF difference is **not** the probability yet:

```
_cdf2(t) = erfc(-t/√2) = 2·Φ(t)
upper - lower = 2 × true bin mass
```

So the value needs a ×0.5 to become the real probability — DCVC just applies that 0.5 *outside* the CDF helper, at the end of `get_prob_train` (`prob = torch.clamp_min(0.5 * prob, 1e-9)`). After that step, the probability is exactly `Φ(v+0.5) - Φ(v-0.5)`, which is the same pmf the coder tabulates in `GaussianEncoder.update()` (`Normal(0, scale).cdf(v ± 0.5)`).

That means our earlier observation was backwards: dropping the `0.5` doesn't fix the estimate, it **under**-charges every y symbol by ~1 bit (`-log2(2p) = -log2(p) - 1`). Keeping the `0.5` gives the true value, so a forward rate that is ~1 bit/symbol higher than the naive `erfc`-difference estimate is expected and correct.

Sorry for the confusion, and thanks again for the careful report!

> I think this is meant to shape the training objective because it directly affects the rate portion of the loss.
> 
> The added bit only affects probabilities that are significantly higher than 1e-9, which should encourage the entropy model to learn a distribution that tends to produce a lower probability, and avoid the added cost. This would lead to a lower entropy in general, and better compression.
> 
> Hopefully we could get a confirmation from the authors, but you could run a test that checks how the entropy during training changes with and without scaling.

**Comment by Zyan20** [comment]
Thanks for the careful write-up — really appreciate it!

We went back through the code and our measurement was mistaken. The key is in `get_prob_train`:

```python
@staticmethod
def get_prob_train(values, scales):
    def _cdf2(inputs):
        const = float(-(2 ** -0.5))
        return torch.erfc(const * inputs)

    dtype = values.dtype
    values = values.float()
    scales = scales.float()
    scales = LowerBound.apply(scales, 0.11)
    values = torch.abs(values)
    upper = _cdf2((0.5 - values) / scales)
    lower = _cdf2((-0.5 - values) / scales)
    prob = upper - lower
    prob = torch.clamp_min(0.5 * prob, 1e-9)
    return prob.to(dtype)
```

The point is that `_cdf2` is **missing a 0.5**: `erfc(-t/√2) = 2·Φ(t)`, so it returns `2Φ`, not `Φ`. Hence `upper - lower` is **2×** the true bin mass, and the `0.5 *` at the end is what makes it the real probability:

```
0.5 * (upper - lower) = Φ(v+0.5) - Φ(v-0.5) = P(v-0.5 < x < v+0.5)
```

which is exactly the pmf the coder tabulates in `GaussianEncoder.update()` (`Normal(0, scale).cdf(v ± 0.5)`).

So our earlier conclusion was backwards: dropping the `0.5` doesn't fix anything, it **under**-charges each y symbol by ~1 bit (`-log2(2p) = -log2(p) - 1`). Keeping the `0.5` gives the true value — so a forward rate ~1 bit/symbol above the raw `erfc`-difference estimate is expected and correct. (The same `0.5` is written *inside* the CDF helper in e.g. CompressAI's `_standardized_cumulative`; DCVC just factors it out to the call site.)

Sorry for the confusion, and thanks again!

> I think this is meant to shape the training objective because it directly affects the rate portion of the loss.
> 
> The added bit only affects probabilities that are significantly higher than 1e-9, which should encourage the entropy model to learn a distribution that tends to produce a lower probability, and avoid the added cost. This would lead to a lower entropy in general, and better compression.
> 
> Hopefully we could get a confirmation from the authors, but you could run a test that checks how the entropy during training changes with and without scaling.
